src/gbxml/Test/ReverseTranslator_GTest.cpp

```diff
@@ -562,3 +496,32 @@ TEST_F(gbXMLFixture, ReverseTranslator_TwoStoryOffice_Trane)
   bool test = forwardTranslator.modelToGbXML(*model, outputPath);
   EXPECT_TRUE(test);
 }
+
+TEST_F(gbXMLFixture, ReverseTranslator_3951_Surface)
+{
+
+  openstudio::path inputPath = resourcesPath() / openstudio::toPath("gbxml/3951_Geometry_bug.xml");
+
+  openstudio::gbxml::ReverseTranslator reverseTranslator;
+  boost::optional<openstudio::model::Model> _model = reverseTranslator.loadModel(inputPath);
+  ASSERT_TRUE(_model);
+
+  // Check all the surfaces that are supposed to be floors and ceilings
+  {
+    auto _surf = _model->getModelObjectByName<Surface>("storey-1-slabongrade-space-1");
+    ASSERT_TRUE(_surf);
+    EXPECT_EQ("Floor", _surf->surfaceType());
+    auto _space = _surf->space();
+    ASSERT_TRUE(_space);
+    EXPECT_EQ("storey-1-space-1", _space->nameString());
+  }
+
+  {
+    auto _surf = _model->getModelObjectByName<Surface>("storey-1-ceiling-space-1");
+    ASSERT_TRUE(_surf);
+    EXPECT_EQ("RoofCeiling", _surf->surfaceType());
+    auto _space = _surf->space();
+    ASSERT_TRUE(_space);
+    EXPECT_EQ("storey-1-space-1", _space->nameString());
+  }
+}
```

Comment:
Added new test with an extract from the user-provided file in #3951: I kept only the bare minimum to make a MCVE